src/sidebar/util/test/copy-to-clipboard-test.js

```diff
@@ -62,33 +57,41 @@ describe('copy-to-clipboard', () => {
   describe('copyPlainText', () => {
     it('writes provided text to clipboard', async () => {
       const text = 'Lorem ipsum dolor sit amet';
-      const navigator = createFakeNavigator();
+      const writeText = sinon.stub();
 
-      await copyPlainText(text, navigator);
+      await copyPlainText(text, createFakeNavigator({ writeText }));
 
-      assert.calledWith(navigator.clipboard.writeText, text);
-      assert.notCalled(navigator.clipboard.write);
+      assert.calledWith(writeText, text);
     });
   });
 
   describe('copyHTML', () => {
     it('writes provided text to clipboard', async () => {
       const text = 'Lorem ipsum dolor sit amet';
-      const navigator = createFakeNavigator();
+      const write = sinon.stub();
 
-      await copyHTML(text, navigator);
+      await copyHTML(text, createFakeNavigator({ write }));
 
-      assert.called(navigator.clipboard.write);
-      assert.notCalled(navigator.clipboard.writeText);
+      assert.called(write);
     });
 
-    it('falls back to plain text if rich text is not supported', async () => {
+    it('falls back to execCommand if clipboard API is not supported', async () => {
       const text = 'Lorem ipsum dolor sit amet';
-      const navigator = createFakeNavigator({ supportsWrite: false });
+      const clipboardData = new DataTransfer();
+      const document = Object.assign(new EventTarget(), {
+        execCommand: sinon.stub().callsFake(command => {
+          if (command === 'copy') {
+            document.dispatchEvent(
+              new ClipboardEvent('copy', { clipboardData }),
+            );
+          }
+        }),
+      });
 
-      await copyHTML(text, navigator);
+      await copyHTML(text, createFakeNavigator({}), document);
 
-      assert.calledWith(navigator.clipboard.writeText, text);
+      assert.calledWith(document.execCommand, 'copy');
+      assert.equal(clipboardData.getData('text/html'), text);
```

Comment:
I tried to test this by not mocking the `document` and adding an actual listener for the copy event, but for some reason it never gets invoked.

I know I have faced this in the past, so I'll try to give it another thought.

Reply:
> I tried to test this by not mocking the document and adding an actual listener for the copy event, but for some reason it never gets invoked.

AFAIK `execCommand('copy')` can only be invoked from the context of a user gesture, so dispatching a fake event is the best you can do under our current testing environment, since we can't simulate a user gesture happening.

Reply:
I thought it might be related with incorrect bubbling/capturing/propagating and was going to test some combinations, but you are probably right.